### PR TITLE
Add double precision support for World Triplanar Mapping

### LINKED ISSUE #91304: Triplanar broke with precision=double
### Tested versions

- Reproducible in https://github.com/V-Sekai/godot/commit/c4733e80030f01182c91f69245a0e5ad247b4e07
- Not reproducible in 4.2

### System information

Mac Mini, Apple Silicon, M2 Pro

### Issue description

Triplanar broke in tests.

image

https://github.com/V-Sekai/godot-double-precision-test

See https://github.com/godotengine/godot/pull/90950#issuecomment-2077844399

### Steps to reproduce

1. Ensure you have a double precision version of Godot Engine
1. Load mrp https://github.com/V-Sekai/godot-double-precision-test
2. Switch to the far version of the test scene.
3. Check that tri-planar is fuzzy on single-precision
3. Check that tri-planar is fuzzy on double-precision (bug)

### Minimal reproduction project (MRP)

https://github.com/V-Sekai/godot-double-precision-test

### PR DESCRIPTION
**Pull request by joao-pedro-braz**
## Description

This PR adds emulated doubles support for the World UV Triplanar Mapping. It also adds some QoL features to both ease the "integration" with the double precision pipeline and also to allow greater flexibility when dealing with Triplanar UVs.

Features:
  - New mat4 TRIPLANAR_MATRIX vertex-only property (the same as MODEL_MATRIX by default).
  - New "hint_triplanar_mat" hint for mat4 uniforms, which allow users to specify a custom TRIPLANAR_MATRIX. Naturally, only a single instance of the hint is allowed per shader.
  - New vec3 TRIPLANAR_POSITION vertex-only property, which holds the VERTEX transformed by the TRIPLANAR_MATRIX (With emulated doubles, if applicable).
  - Updated Standard 3D Material to make use of the above features.
  - Updated VisualShader to include the TRIPLANAR_MATRIX and TRIPLANAR_POSITION as valid Inputs.
  - Added a "hint_triplanar_enabled" toggle to the VisualShaderNodeTransformParameter to allow the new "hint_triplanar_mat" hint to be set.

Below is a Demo which was tailored to showcase these additions.
Video:

https://user-images.githubusercontent.com/37230465/229313783-81fd6a40-4578-41b4-9e57-9551d86d2698.mp4

Project: [TestDoubles.zip](https://github.com/godotengine/godot/files/11130559/TestDoubles.zip)

Note: The shader related changes were applied to all renderers (Forward+, Mobile and Compatibility).

Note²: Given the lack of emulated doubles support for the Compatibility renderer, only the single precision path was implemented.

Note³: Related to https://github.com/godotengine/godot/pull/75462, where the need for this was brought up by @Zylann (specifically the support for custom matrices when computing the TRIPLANAR_POSITION).

- *Production edit: This closes https://github.com/godotengine/godot/issues/91304.*

### CODE REVIEW

**Comment by joao-pedro-braz** [comment]
(forgot to --doctool, sorry about that)

**Comment by Chaosus** [comment]
Would you like to change the existed `VisualShaderNodeTextureParameterTriplanar` to support this by default or (if its break compatibility) by setting a flag?

**Comment by joao-pedro-braz** [comment]
That's actually a great idea!
A simple "World Triplanar" toggle should suffice, I'll have a go at it rn

**Comment by joao-pedro-braz** [comment]
image
So much simpler

**Comment by joao-pedro-braz** [comment]
I think I ran --doctool with a double precision build 😒

**Review by Calinou — COMMENTED** [review]
Tested locally, it mostly works as expected. World triplanar now works correctly on still objects (even if the camera moves), but moving the object when far away from the world origin will have visible snapping in the final UV.

**Testing project:** [test_triplanar_double.zip](https://github.com/godotengine/godot/files/13065014/test_triplanar_double.zip)

All videos below taken at X = 5,000,000. The object uses a StandardMaterial3D and is moved using a script.

### `master`, single precision

https://github.com/godotengine/godot/assets/180032/82e3380f-e4a9-4d4a-b3c5-383942e75d28

### `master`, double precision

https://github.com/godotengine/godot/assets/180032/c1fec71e-e522-4be2-82c0-7f9d66ea8bed

### *This PR*, double precision

https://github.com/godotengine/godot/assets/180032/f4c30018-482f-430c-ad30-e9d284968e35

**Review by fire — APPROVED** [review]
The pull request is a good idea but needs a rebase. Calinou tested on the last version.

**Comment by joao-pedro-braz** [comment]
> The pull request is a good idea but needs a rebase. Calinou tested on the last version.

Rebased.
Thanks!

**Comment by fire** [comment]
There are some rebases needed, I'll try to see if anything's blocking this for 4.3.

File: scene/resources/visual_shader_nodes.cpp [hunk 1]
```diff
@@ -6191,6 +6200,9 @@ Transform3D VisualShaderNodeTransformParameter::get_default_value() const {
 
 String VisualShaderNodeTransformParameter::generate_global(Shader::Mode p_mode, VisualShader::Type p_type, int p_id) const {
 	String code = _get_qual_str() + "uniform mat4 " + get_parameter_name();
+	if (hint_triplanar_enabled) {
+		code += ":hint_triplanar_mat";
```
Comment (AThousandShips) [inline scene/resources/visual_shader_nodes.cpp:6204]:
```suggestion
		code += " : hint_triplanar_mat";
```

**Review by AThousandShips — APPROVED** [review]

**Comment by fire** [comment]
I did a push to salvage the pr.

image

Changes here.

**Comment by fire** [comment]
Superseded by: https://github.com/godotengine/godot/pull/91380

**Comment by Calinou** [comment]
Superseded by https://github.com/godotengine/godot/pull/91380. Thanks for the contribution nonetheless!

**Pull request closed without merging**